### ISSUE TITLE
Oracle GraalVM 20.0.1+9.1 crashed on macOS 13.4 Ventura (M1 processor)

### ISSUE THREAD

**Issue by edgebroker** [oracle/graal#6796]
**Describe the issue**
JVM crashed.

**Steps to reproduce the issue**
I tested with Flow Director (flowdirector.io) and the JVM crashed. I don't know whether I can reproduce it.

**Describe GraalVM and your environment:**
 - GraalVM version (latest snapshot builds can be found [here](https://github.com/graalvm/graalvm-ce-dev-builds/releases)), or commit id if built from source: **[e.g. EE 19.3, CE 20.0, CE 20.1.0-dev-20200304_0848]**
 - JDK major version: **Oracle GraalVM 20.0.1+9.1**
 - OS: **macOS 13.4 Ventura**
 - Architecture: **Apple M1**

**More details**
Add any other information about the problem here. Especially important are stack traces or log output. Feel free to link to gists or to screenshots if necessary.
```
#
# A fatal error has been detected by the Java Runtime Environment:
#
#  SIGSEGV (0xb) at pc=0x00000001061c1280, pid=27574, tid=60419
#
# JRE version: Java(TM) SE Runtime Environment Oracle GraalVM 20.0.1+9.1 (20.0.1+9) (build 20.0.1+9-jvmci-23.0-b12)
# Java VM: Java HotSpot(TM) 64-Bit Server VM Oracle GraalVM 20.0.1+9.1 (20.0.1+9-jvmci-23.0-b12, mixed mode, sharing, tiered, jvmci, jvmci compiler, compressed oops, compressed class ptrs, g1 gc, bsd-aarch64)
# Problematic frame:
# V  [libjvm.dylib+0x881280]  SharedRuntime::resolve_sub_helper_internal(methodHandle, frame const&, CompiledMethod*, bool, bool, Handle, CallInfo&, Bytecodes::Code, JavaThread*)+0xd4
#
# No core dump will be written. Core dumps have been disabled. To enable core dumping, try "ulimit -c unlimited" before starting Java again
#
# An error report file with more information is saved as:
# /Users/am/Downloads/testconnect/flowdirector/scripts/hs_err_pid27574.log
[330.574s][warning][os] Loading hsdis library failed
#
# If you would like to submit a bug report, please visit:
#   https://bugreport.java.com/bugreport/crash.jsp
#
```
[hs_err_pid27574.log](https://github.com/oracle/graal/files/11746794/hs_err_pid27574.log)

**Comment by edgebroker** [comment]
It seems it is reproducible. Attached are 2 more error files.
[hs_err_pid27775.log](https://github.com/oracle/graal/files/11747481/hs_err_pid27775.log)
[hs_err_pid28001.log](https://github.com/oracle/graal/files/11747484/hs_err_pid28001.log)

**Comment by tkrodriguez** [comment]
Disassembly from libjvm.dylib looks like this:
```
libjvm.dylib`SharedRuntime::resolve_sub_helper_internal:
<+0>:   sub    sp, sp, #0xd0
<+4>:   stp    x28, x27, [sp, #0x70]
<+8>:   stp    x26, x25, [sp, #0x80]
<+12>:  stp    x24, x23, [sp, #0x90]
<+16>:  stp    x22, x21, [sp, #0xa0]
<+20>:  stp    x20, x19, [sp, #0xb0]
<+24>:  stp    x29, x30, [sp, #0xc0]
<+28>:  add    x29, sp, #0xc0
<+32>:  mov    x23, x7
<+36>:  mov    x22, x6
<+40>:  mov    x28, x5
<+44>:  mov    x27, x4
<+48>:  mov    x24, x3
<+52>:  mov    x20, x2
<+56>:  mov    x21, x0
<+60>:  add    x25, sp, #0x50
<+64>:  stp    xzr, xzr, [sp, #0x58]
<+68>:  stp    xzr, xzr, [sp, #0x38]
<+72>:  str    wzr, [sp, #0x48]
<+76>:  ldr    x8, [x0]
<+80>:  add    x8, x8, #0x48
<+84>:  ldapr  x19, [x8]
<+88>:  str    x1, [sp, #0x8]
<+92>:  cbz    x19, 0x881228             ; <+124> at sharedRuntime.cpp
<+96>:  ldr    x8, [x19]
<+100>: ldr    x8, [x8, #0x98]
<+104>: mov    x0, x19
<+108>: blr    x8
<+112>: cmp    w0, #0x0
<+116>: csel   x26, x19, xzr, ne
<+120>: b      0x88122c                  ; <+128> at sharedRuntime.cpp
<+124>: mov    x26, #0x0
<+128>: add    x8, x25, #0x8
<+132>: str    x8, [sp, #0x10]
<+136>: ldr    x8, [x20]
<+140>: ldr    x8, [x8, #0x10]
<+144>: mov    x0, x20
<+148>: blr    x8
<+152>: mov    x25, x0
<+156>: cbz    w24, 0x88126c             ; <+192> at sharedRuntime.cpp:1394:5
<+160>: ldr    x19, [x29, #0x10]
<+164>: ldr    x0, [x22, #0x8]
<+168>: bl     0x793b38                  ; Method::can_be_statically_bound at method.cpp:820
<+172>: mov    x3, x0
<+176>: cmp    w23, #0xe9
<+180>: b.ne   0x881280                  ; <+212> [inlined] Handle::non_null_obj() const at handles.hpp:71:109
<+184>: mov    x1, #0x0
<+188>: b      0x8812b8                  ; <+268> at sharedRuntime.cpp:1389:5
<+192>: add    x2, sp, #0x50
<+196>: mov    x0, x21
<+200>: mov    x1, x25
<+204>: bl     0x2ab3d8                  ; CompiledStaticCall::compute_entry at compiledIC.cpp:633
<+208>: b      0x8812ec                  ; <+320> at sharedRuntime.cpp:1399:22
<+212>: ldr    x8, [x28]
<+216>: adrp   x9, 854
<+220>: add    x9, x9, #0x71             ; UseCompressedClassPointers
<+224>: ldrb   w9, [x9]
<+228>: cbz    w9, 0x8812b4              ; <+264> [inlined] oopDesc::klass() const + 48 at oop.inline.hpp:88:22
<+232>: ldr    w8, [x8, #0x8]
<+236>: adrp   x9, 801
<+240>: add    x9, x9, #0x8c8            ; CompressedKlassPointers::_narrow_klass
<+244>: ldr    x10, [x9]
<+248>: ldr    w9, [x9, #0x8]
<+252>: lsl    x8, x8, x9
<+256>: add    x1, x8, x10
<+260>: b      0x8812b8                  ; <+268> at sharedRuntime.cpp:1389:5
<+264>: ldr    x1, [x8, #0x8]
<+268>: add    x5, sp, #0x38
<+272>: mov    x0, x21
<+276>: mov    x2, x27
<+280>: mov    x4, x25
<+284>: mov    x6, x19
<+288>: bl     0x2ab05c                  ; CompiledIC::compute_monomorphic_entry at compiledIC.cpp:506
<+292>: ldr    x8, [x19, #0x8]
<+296>: cbz    x8, 0x8812ec              ; <+320> at sharedRuntime.cpp:1399:22
<+300>: mov    w19, #0x0
<+304>: ldr    x21, [sp, #0x10]
<+308>: ldrb   w8, [sp, #0x4b]
<+312>: cbnz   w8, 0x881448              ; <+668> [inlined] CompiledICInfo::~CompiledICInfo() + 8 at compiledIC.hpp:137:55
<+316>: b      0x881454                  ; <+680> [inlined] StaticCallInfo::~StaticCallInfo() at compiledIC.hpp:324:7
<+320>: add    x0, sp, #0x18
<+324>: mov    x1, x20
<+328>: bl     0x2aa208                  ; CompiledICLocker::CompiledICLocker at compiledIC.cpp:57
<+332>: ldr    x8, [x21]
<+336>: ldrb   w8, [x8, #0x2a]
<+340>: tbnz   w8, #0x0, 0x881330        ; <+388> at sharedRuntime.cpp:1391:22
<+344>: cbz    x26, 0x881340             ; <+404> at sharedRuntime.cpp:1420:11
<+348>: ldr    x8, [x26]
<+352>: ldr    x8, [x8, #0x98]
<+356>: mov    x0, x26
<+360>: blr    x8
<+364>: cbz    w0, 0x881330              ; <+388> at sharedRuntime.cpp:1391:22
<+368>: ldr    x8, [x21]
<+372>: add    x8, x8, #0x48
<+376>: ldapr  x8, [x8]
<+380>: cmp    x26, x8
<+384>: b.eq   0x881340                  ; <+404> at sharedRuntime.cpp:1420:11
<+388>: mov    w19, #0x0
<+392>: mov    w20, #0x1
<+396>: ldr    x21, [sp, #0x10]
<+400>: b      0x881434                  ; <+648> at sharedRuntime.cpp:1441:3
<+404>: cbz    w24, 0x88139c             ; <+496> at sharedRuntime.cpp:1428:59
<+408>: ldr    x8, [sp, #0x8]
<+412>: ldr    x21, [x8, #0x8]
<+416>: mov    w0, #0x20
<+420>: mov    w1, #0x0
<+424>: bl     0x86aa7c                  ; resource_allocate_bytes at resourceArea.cpp:66
<+428>: mov    x19, x0
<+432>: cbz    x0, 0x881370              ; <+452> at compiledIC.hpp
<+436>: sub    x2, x21, #0x4
<+440>: mov    x0, x19
<+444>: mov    x1, x20
<+448>: bl     0x2aa8d0                  ; CompiledIC::CompiledIC at compiledIC.cpp:211
<+452>: ldr    x21, [sp, #0x10]
<+456>: mov    x0, x19
<+460>: bl     0x2aadec                  ; CompiledIC::is_clean at compiledIC.cpp:396
<+464>: cbz    w0, 0x88142c              ; <+640> at sharedRuntime.cpp:1391:22
<+468>: add    x1, sp, #0x38
<+472>: mov    x0, x19
<+476>: bl     0x2aae5c                  ; CompiledIC::set_to_monomorphic at compiledIC.cpp:405
<+480>: tbnz   w0, #0x0, 0x88142c        ; <+640> at sharedRuntime.cpp:1391:22
<+484>: mov    w20, #0x0
<+488>: mov    w19, #0x0
<+492>: b      0x881434                  ; <+648> at sharedRuntime.cpp:1441:3
<+496>: cmp    w23, #0xb8
<+500>: b.ne   0x8813d0                  ; <+548> at sharedRuntime.cpp
<+504>: ldr    x0, [x21]
<+508>: bl     0x794064                  ; Method::needs_clinit_barrier at method.cpp:914
<+512>: cbz    x26, 0x8813d0             ; <+548> at sharedRuntime.cpp
<+516>: cbz    w0, 0x8813d0              ; <+548> at sharedRuntime.cpp
<+520>: ldrb   w8, [x26, #0x5e]
<+524>: cmp    w8, #0x3
<+528>: b.ne   0x8813d0                  ; <+548> at sharedRuntime.cpp
<+532>: mov    w20, #0x0
<+536>: mov    w19, #0x1
<+540>: ldr    x21, [sp, #0x10]
<+544>: b      0x881434                  ; <+648> at sharedRuntime.cpp:1441:3
<+548>: ldr    x8, [sp, #0x8]
<+552>: ldr    x1, [x8, #0x8]
<+556>: ldr    x8, [x20]
<+560>: ldr    x8, [x8, #0x1d8]
<+564>: mov    x0, x20
<+568>: blr    x8
<+572>: mov    x19, x0
<+576>: cbz    w25, 0x881410             ; <+612> at sharedRuntime.cpp:1438:18
<+580>: ldr    x8, [x20, #0x68]
<+584>: ldrh   w8, [x8, #0x30]
<+588>: cmp    w8, #0xc1
<+592>: b.ne   0x881410                  ; <+612> at sharedRuntime.cpp:1438:18
<+596>: add    x2, sp, #0x50
<+600>: mov    x0, x19
<+604>: mov    x1, x21
<+608>: bl     0x2ab470                  ; CompiledStaticCall::compute_entry_for_continuation_entry at compiledIC.cpp:648
<+612>: mov    x0, x19
<+616>: bl     0x2ab260                  ; CompiledStaticCall::is_clean at compiledIC.cpp:587
<+620>: ldr    x21, [sp, #0x10]
<+624>: cbz    w0, 0x88142c              ; <+640> at sharedRuntime.cpp:1391:22
<+628>: add    x1, sp, #0x50
<+632>: mov    x0, x19
<+636>: bl     0x2ab35c                  ; CompiledStaticCall::set at compiledIC.cpp:615
<+640>: mov    w19, #0x0
<+644>: mov    w20, #0x1
<+648>: add    x0, sp, #0x18
<+652>: bl     0x2aa284                  ; CompiledICLocker::~CompiledICLocker at compiledIC.cpp:60
<+656>: orr    w19, w20, w19
<+660>: ldrb   w8, [sp, #0x4b]
<+664>: cbz    w8, 0x881454              ; <+680> [inlined] StaticCallInfo::~StaticCallInfo() at compiledIC.hpp:324:7
<+668>: ldr    x0, [sp, #0x40]
<+672>: cbz    x0, 0x881454              ; <+680> [inlined] StaticCallInfo::~StaticCallInfo() at compiledIC.hpp:324:7
<+676>: bl     0x11b1b0                  ; FreeHeap at allocation.cpp:67
<+680>: mov    x0, x21
<+684>: bl     0x440558                  ; methodHandle::~methodHandle at handles.cpp:97
<+688>: mov    x0, x19
<+692>: ldp    x29, x30, [sp, #0xc0]
<+696>: ldp    x20, x19, [sp, #0xb0]
<+700>: ldp    x22, x21, [sp, #0xa0]
<+704>: ldp    x24, x23, [sp, #0x90]
<+708>: ldp    x26, x25, [sp, #0x80]
<+712>: ldp    x28, x27, [sp, #0x70]
<+716>: add    sp, sp, #0xd0
<+720>: ret
```
We're crashing a +0xd4 which is +212:
```
<+212>: ldr    x8, [x28]
```
x28 appears to be a copy of x5 which is the sixth incoming argument and is supposed to be a Handle to the receiver.  If the call site requires a receiver and it's actually null then find_callee_info should have thrown an NPE.  Since we are in RuntimeStub::resolve_opt_virtual_call we should have a receiver so I'm not sure how we get to this code with a null receiver.

**Comment by tkrodriguez** [comment]
It doesn't seem like it can be that simple though, otherwise I think we'd be crashing a lot more.  Being able to reproduce it would be a great help.

**Comment by edgebroker** [comment]
I ran it with Oracle GraalVM 17.0.7 and it crashes, too.

[hs_err_pid37367.log](https://github.com/oracle/graal/files/11756758/hs_err_pid37367.log)

**Comment by edgebroker** [comment]
Well, it doesn't work with 23.0 CE anymore as the `js.nashorn-compat` seems to be gone?

**Comment by edgebroker** [comment]
You can test it with Oracle GraalVM (install graal.js). It just needs to be in your `PATH` with `JAVA_HOME` orderly set. Then execute `./startserver` from the `scripts` directory. 

When it comes up, connect to `localhost:8080` & install the _Warehouse Sample App_ from the install wizard displayed. 

Then just wait. Takes several minutes until it happens because the Warehouse App starts slowly. You can watch the Warehouse by logging into it (User/password: `admin` / `changeme`).

**Comment by edgebroker** [comment]
> Could you provide some more detail about how to run this? If I download flowdirector-4.1.0.tar.gz from https://www.flowdirector.io/start/ and run it with 23.0 it dies like this during startup.
> 
> ```
> java.lang.IllegalArgumentException: Could not find option with name js.nashorn-compat.
> 	at org.graalvm.truffle/com.oracle.truffle.polyglot.PolyglotEngineException.illegalArgument(PolyglotEngineException.java:131)
> 	at org.graalvm.truffle/com.oracle.truffle.polyglot.OptionValuesImpl.failNotFound(OptionValuesImpl.java:288)
> 	at org.graalvm.truffle/com.oracle.truffle.polyglot.PolyglotEngineImpl.parseOptions(PolyglotEngineImpl.java:763)
> 	at org.graalvm.truffle/com.oracle.truffle.polyglot.PolyglotEngineImpl.<init>(PolyglotEngineImpl.java:322)
> 	at org.graalvm.truffle/com.oracle.truffle.polyglot.PolyglotImpl.buildEngine(PolyglotImpl.java:283)
> 	at com.oracle.graal.graal_enterprise/com.oracle.truffle.polyglot.enterprise.EnterprisePolyglotImpl.buildEngine(stripped:136)
> 	at org.graalvm.sdk/org.graalvm.polyglot.Engine$Builder.build(Engine.java:681)
> ```

There is currently an issue installing `graal.js`:

`bin/gu install js
Downloading: Component catalog from www.graalvm.org
Processing Component: Graal.js
Processing Component: ICU4J
Processing Component: TRegex
Additional Components are required:
    ICU4J (org.graalvm.icu4j, version 23.0.0), required by: Graal.js (org.graalvm.js)
    TRegex (org.graalvm.regex, version 23.0.0), required by: Graal.js (org.graalvm.js)
Downloading: Component js: Graal.js from github.com
I/O error occurred: https://github.com/graalvm/graalvm-ce-builds/releases/download/graal-23.0.0/js-installable-svm-java20-darwin-aarch64-23.0.0.jar`

I've reported that via Slack and they are working on it. Once fixed, you can install graal.js and retest.

**Comment by edgebroker** [comment]
I was using the Oracle GraalVM distros for Java 17 & 20 and both produce the crash. I had no problems installing graal.js with them. I tried to use the GraalVM CE distro 23.0 but had the I/O error installing graal.js. I reported that via Slack and it seems to be a known issue (the I/O error).

**Comment by tkrodriguez** [comment]
So it's been running for a while without crashing.  I'm going to try relaunching in periodically in case it's something less reliable on my setup.  Maybe there's something in your local setup that's different?  Does it reproduce for you from a clean download of flowdirector-4.1.0.tar.gz?  I'm running on an m1 pro with ventura so I think I'm pretty close hardware wise.

**Comment by edgebroker** [comment]
I have a bunch of apps installed but thought (from the error file stack trace) it might be the Warehouse App. I tested it with 2 apps on the production version but it did not happen.

So here is my FD installation that crashes:

https://www.dropbox.com/s/jwsq1xjeovtog0w/flowdirector-crash.tar.gz

It's a 4.1.0 with apps installed. Please unpack and run 

`./install
`

which installs Oracle GraalVM Java 20. I have modified the install script accordingly. 

Then 

`./startserver
`

and let it run for about 5 minutes. Then login to the Warehouse App. It immediately crashed right after login.

**Comment by tkrodriguez** [comment]
That reproduces it for me.  I can see that it's a mismatch between the is_virtual flag and what find_callee_info thinks is going on.  It thinks says we're at an invokestatic which obviously doesn't have a receiver so we crash later because receiver is null.

**Comment by tkrodriguez** [comment]
This seems to be a bad interaction between our method handle invoke machinery and the macro nodes we use for the `StringBuilder` optimization. Method handle invokes capture some extra metadata in a `ResolvedMethodHandleCallTargetNode` which is required when the target of a method handle doesn't end up being inlined.  The StringBuilder optimization captures calls to `StringBuilder.append` as `MacroNodes` so it can perform high level transformations and then they are later reconsidered for inlining if they aren't optimized.  We are losing the extra method handle information because MacroNode doesn't hang onto the `ResolvedMethodHandleCallTargetNode`.  Any `MacroNode` is exposed to this problem.  This particular case can be worked around by disabling the `StringBuilder` optimization using `-Dgraal.OptStringConcat=false`.  One thing I'm not clear on is why this only happens in 23 as none of the relevant machinery has really changed but I think a fix should be straightforward.  Thanks for the report!

**Comment by edgebroker** [comment]
Great you found it! And that quick! Highly appreciate it. I'll add that option to our `startserver` scripts. 

Thanks very much!

**Comment by tkrodriguez** [comment]
This should be fixed by https://github.com/oracle/graal/commit/c78e42760217c86365b16c90b84fa53ba21e0a2b

**Comment by edgebroker** [comment]
In which release will this fix be included?

**Comment by tkrodriguez** [comment]
The fix is in 23.1 which is the very latest JDK21 release  https://www.graalvm.org/release-notes/JDK_21
